### PR TITLE
Add test to reproduce behaviour

### PR DESCRIPTION
**Pull request by sharpner**
#1275 

We are totally brainfucked, but this should reproduce our confusing somewhat.

### CODE REVIEW

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        $container->add($itemOne);
+        $container->add($itemTwo);
+        $container->add($itemThree);
+
+        $this->assertSame(
+            array('Number One', 'Number Two', 'Number Three'),
+            array_map($getNameCallback, $container->items->toArray())
+        );
+
+        $this->dm->refresh($container);
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php]: This will revert the changes made to ```$container``` in L75..L77 since they haven't been flushed to the database.

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+        $this->assertSame(
+            array('one', 'two', 'three'),
+            array_map($getNameCallback, $container->$strategy->toArray())
+        );
+
+        $two = $container->$strategy->get(1);
+        $three = $container->$strategy->get(2);
+        $container->items->set(1, $three);
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php]: I lost some hair due to `$container->items` here :D

Comment (gruberro) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php]: Oh damn. I've reviewed the code so often... I'm so sorry :-(

Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1275Test.php]: No biggie, don't worry ;)

**Comment by gruberro** [comment]
Great... I was able to reproduce the error I'm getting within our application by writing a failing but broken test case. I'll invest some time tomorrow to debug our application stack as the error must be somewhere there. Maybe some broken event subscriber or something similar - will see. @malarzm: do you wan't me to remove the test case?

**Comment by malarzm** [comment]
@gruberro I've included all tests from here in #1278 so whatever suits you :)

**Comment by gruberro** [comment]
I finally found the "making it buggy" trigger (https://github.com/doctrine/mongodb-odm/commit/ab89a77dd3a50707e348ba26b2e8235cc9acd629). By using a ```preUpdate``` lifecycle callback (even if it's doing nothing) elements are removed from the collection. I lost some hair either in the meantime ;)

**Comment by malarzm** [comment]
I suppose that'll be because UnitOfWork recomputes change sets for document that had `preUpdate` callback - I remember that the same trigger was reported when reusing embedded document that was already scheduled for orphan removal.

**Comment by sharpner** [comment]
So is this a bug or expected behaviour?

**Comment by malarzm** [comment]
Unfortunate but expected (as of now) behaviour I'd say.

**Comment by sharpner** [comment]
Even though it loses data? with 1.0.2 it works, and with 1.0.3 we lose items...

**Comment by malarzm** [comment]
@alcaeus wrote full explaination on why is this happening: https://github.com/doctrine/mongodb-odm/issues/1275#issuecomment-156501140. If we don't merge #1278 for whatever reasons (I for one don't like making it O(n)) you may add items in new collection and replace the old one entirely.

**Comment by alcaeus** [comment]
@sharpner @gruberro Closing this to consolidate discussion into #1278.

We'll take a look at what exactly is going on with preUpdate and see if we can work around it.
